wasm_utils/src/memory_serialization.rs

```diff
@@ -6,36 +6,72 @@ use serde::{Deserialize, Serialize};
 use serde_json;
 use std::{ffi::CStr, os::raw::c_char, slice};
 
-// TODO #486 - load and store string from wasm memory
-//pub fn load_string(encoded_allocation: u32) -> Result<String, String> {
-//    let maybe_allocation = decode_encoded_allocation(encoded_allocation);
-//    match maybe_allocation {
-//        Err(return_code) => Err(return_code.to_string()),
-//        Ok(allocation) => Ok(load_string_from_allocation(&allocation)),
-//    }
-//}
-///// Write a string in wasm memory
-//pub fn store_string(stack: &mut SinglePageStack, s: &str)
-//    -> Result<SinglePageAllocation, RibosomeErrorCode> {
-//    let mut bytes: Vec<_> = s.to_string().into_bytes();
-//    bytes.push(0); // Add string terminate character (important)
-//    let len = bytes.len();
-//    if len > <u16>::max_value() as usize {
-//        return Err(RibosomeErrorCode::OutOfMemory);
-//    }
-//    return write_in_wasm_memory(stack, &bytes, len as u16);
-//}
+//-------------------------------------------------------------------------------------------------
+// Raw
+//-------------------------------------------------------------------------------------------------
 
-/// Expecting to retrieve a struct from an encoded allocation, but return error string in case of error
-pub fn load_json<'s, T: Deserialize<'s>>(encoded_allocation: u32) -> Result<T, String> {
+/// Convert a string stored in wasm memory into a String.
+fn load_str_from_raw<'a>(ptr_data: *mut c_char) -> &'a str {
+    let ptr_safe_c_str = unsafe { CStr::from_ptr(ptr_data) };
+    let stored_str = ptr_safe_c_str.to_str().unwrap();
+    stored_str
+}
+
+/// Write in wasm memory according to stack state.
+fn write_in_wasm_memory(
+    stack: &mut SinglePageStack,
+    bytes: &Vec<u8>,
+    len: u16,
+) -> Result<SinglePageAllocation, RibosomeErrorCode> {
+    if len as u32 + stack.top() as u32 > U16_MAX {
+        return Err(RibosomeErrorCode::OutOfMemory);
+    }
+    let ptr = stack.allocate(len) as *mut c_char;
+    let ptr_safe = unsafe { slice::from_raw_parts_mut(ptr, len as usize) };
+    for (i, byte) in bytes.iter().enumerate() {
+        ptr_safe[i] = *byte as i8;
+    }
+    SinglePageAllocation::new(ptr as u16, len)
+}
+
+//-------------------------------------------------------------------------------------------------
+// String
+//-------------------------------------------------------------------------------------------------
+
+/// Write a string in wasm memory according to stack state.
+pub fn store_string(
+    stack: &mut SinglePageStack,
+    s: &str,
+) -> Result<SinglePageAllocation, RibosomeErrorCode> {
+    let bytes = s.to_string().into_bytes();
```

Comment:
https://doc.rust-lang.org/1.4.0/std/primitive.str.html#method.as_bytes

Maybe that works better and doesn't copy